modules/standard/GMP.chpl

```diff
@@ -560,7 +560,7 @@ module GMP {
     }
     proc set_str(str:string, base:int=0)
     {
-      on this do mpz_set_str(this.mpz, str.c_str(), base.safeCast(c_int));
+      on this do mpz_set_str(this.mpz, str.localize().c_str(), base.safeCast(c_int));
```

Comment:
Unfortunately, a helper method along those lines can't exist without leaking the string. If you create the string object inside the helper, returning the c_stri() of that will point to freed memory after the function returns and the string goes out of scope. Making the buffer persist out of the function will cause it to be leaked.
________________________________________
From: Brad Chamberlain [notifications@github.com]
Sent: Wednesday, December 02, 2015 10:07 AM
To: chapel-lang/chapel
Cc: Kyle Brady
Subject: Re: [chapel] Add record-based strings to Chapel (#2884)

In modules/standard/GMP.chpl<https://github.com/chapel-lang/chapel/pull/2884#discussion_r46450607>:

> @@ -560,7 +560,7 @@ module GMP {
>      }
>      proc set_str(str:string, base:int=0)
>      {
> -      on this do mpz_set_str(this.mpz, str.c_str(), base.safeCast(c_int));
> +      on this do mpz_set_str(this.mpz, str.localize().c_str(), base.safeCast(c_int));


To clarify, I don't object to the calls, but wanted to make sure I understood the current and proposed behavior (and continue to think that the difference is worth elevating as part of "awareness of what changes the string patch brings" since people writing such code in the future will need to take an action like this.

While I think I agree that c_str() shouldn't automatically localize, I could imagine a helper method that combined localize + c_str() as a helper method, given that it looks likely to be a common pattern, at least for internal devs. But I'm not suggesting we do that before merging.

�
Reply to this email directly or view it on GitHub<https://github.com/chapel-lang/chapel/pull/2884/files#r46450607>.